### PR TITLE
Enable xarray `.chunk()` style input for target chunks

### LINKED ISSUE #83: Specify chunk plan for all variables via coordinates?
When rechunking with tools like Unidata's [nccopy](https://www.unidata.ucar.edu/software/netcdf/workshops/2011/utilities/Nccopy.html), we can specify the desired chunk sizes for each coordinate, and then all variables that depend on those coordinates will be chunked accordingly, for example:
```
nccopy -d1 -c time/10,lon/20,lat/30 netcdf.nc netcdf_chunked.nc
```
Is it possible to accomplish the same with rechunker?

The docs seem to indicate that you need to provide a dictionary that contains every variable that you want to chunk.  
I'm currently working on a dataset with 180 data variables!

### PR DESCRIPTION
**Pull request by jbusecke**
- [x] Closes #83

I have started implementing this functionality (many thanks to @pmav99 for the help over in #83)

I think the main functionality works, but I have a few outstanding questions I would like to discuss. 

The current implementation will rechunk to the full length of a dimension if the dimension is not provided in `target_chunks`: 

If ds has two dimensions: x/y with lengths 10/20 and chunks 5/2

Then what do we want to happen to dimension `y` if we provide `target_chunks={'x':5}`? @pmav99 s implementation currently defaults to the length of y as chunksize for `y`. 

I would say we should stay as close to the implementation of xarrays `.chunk()`, which I believe just leaves the chunks alone. Thoughts?

Also, should this only work for datasets or also dataarrays?

### CODE REVIEW

File: tests/test_rechunk.py [hunk 1]
```diff
@@ -361,7 +384,11 @@ def rechunk_args(tmp_path, request):
         target_chunks = (8000, 200)
 
         args.update(
-            {"source": array, "target_chunks": target_chunks, "max_mem": max_mem,}
+            {
+                "source": array,
+                "target_chunks": target_chunks,
+                "max_mem": max_mem,
+            }
```
Comment (pmav99) [inline tests/test_rechunk.py:391]: Is this necessary? it seems to be just formatting

File: tests/test_rechunk.py [hunk 2]
```diff
@@ -207,7 +224,13 @@ def test_rechunk_array(
...
 @pytest.mark.parametrize("max_mem", [25600000])
 @pytest.mark.parametrize(
-    "target_chunks", [(200, 8000), (800, 8000), (8000, 200), (400, 8000),],
+    "target_chunks",
+    [
+        (200, 8000),
+        (800, 8000),
+        (8000, 200),
+        (400, 8000),
+    ],
```
Comment (pmav99) [inline tests/test_rechunk.py:233]: Is this necessary? This seems to be just formatting

File: tests/test_rechunk.py [hunk 3]
```diff
@@ -119,14 +126,24 @@ def test_rechunk_dataset(
 
     # Validate decoded variables
     dst = xarray.open_zarr(target_store, decode_cf=True)
-    target_chunks_expected = (
-        target_chunks["a"]
-        if isinstance(target_chunks["a"], tuple)
-        else (target_chunks["a"]["x"], target_chunks["a"]["y"])
-    )
+    if chunk_by_dims:
```
Comment (pmav99) [inline tests/test_rechunk.py:129]: IMHV having conditionals in tests is usually a good sign that the single test function should be split in two separate ones. This usually helps with readability and makes it easier to understand the problem when the tests break. Obviously you do have some code duplication when you go down this line

File: rechunker/api.py [hunk 4]
```diff
@@ -311,6 +312,30 @@ def rechunk(
...
+    for var in ds.variables:
+        for dim in ds[var].dims:
+            if dim in target_chunks.keys() and target_chunks[dim] <= len(ds[dim]):
+                group_chunks[var].append(target_chunks[dim])
+            else:
+                group_chunks[var].append(len(ds[dim]))
+
+    # rechunk() expects chunks values to be a tuple. So let's convert them
+    group_chunks_tuples = {var: tuple(chunks) for (var, chunks) in group_chunks.items()}
+    return group_chunks_tuples
```
Comment (pmav99) [inline rechunker/api.py:336]: Adding a unit test for `parse_target_chunks_from_dim_chunks()`  wouldn't be a bad idea.

File: rechunker/api.py [hunk 5]
```diff
@@ -311,6 +312,30 @@ def rechunk(
...
 
+def parse_target_chunks_from_dim_chunks(ds, target_chunks):
+    """
+    Calculate ``target_chunks`` suitable for ``rechunker.rechunk()`` using chunks defined for
+    dataset dimensions (similar to xarray's ``.rechunk()``) .
+
+    - If a dimension is missing from ``target_chunks`` then use the full length from ``ds``.
+    - If a chunk in ``target_chunks`` is larger than the full length of the variable in ``ds``,
+      then, again, use the full length from the dataset.
+    - If a dimension chunk is specified as -1, again, use the full length from the dataset.
```
Comment (pmav99) [inline rechunker/api.py:323]: I may be missing something here, but is `-1` handled in the current implementation?

**Review by pmav99 — CHANGES_REQUESTED** [review]
Nice work! I left a few minor remarks. 

BTW, shouldn't the docs of `rechunk()` be updated as well?

File: rechunker/api.py (hunk 5 above)
Comment (jbusecke) [reply rechunker/api.py:323]: It is in [xarrays `chunk`](https://xarray.pydata.org/en/v0.14.0/generated/xarray.DataArray.chunk.html), it is however not super well documented it seems.

File: rechunker/api.py (hunk 4 above)
Comment (jbusecke) [reply rechunker/api.py:336]: Yes, absolutely. I would adapt your test for this from #83.

File: tests/test_rechunk.py (hunk 3 above)
Comment (jbusecke) [reply tests/test_rechunk.py:129]: I did think about that, and generally agree. In order to make that clean I could factor out some of the setup. I originally ended up with tons of duplicated code.

File: tests/test_rechunk.py (hunk 2 above)
Comment (jbusecke) [reply tests/test_rechunk.py:233]: Um yes, that seems to be some automatic formatting. I will make sure to properly reformat for the final PR.

File: tests/test_rechunk.py (hunk 3 above)
Comment (jbusecke) [reply tests/test_rechunk.py]: I did add an additional test to separate the test cases.

File: rechunker/api.py (hunk 4 above)
Comment (jbusecke) [reply rechunker/api.py:336]: Added the test

**Comment by jbusecke** [comment]
I am having two issues here, and would appreciate some input. 

1) I have installed pre-commit but the formatting does not seem to be identical. Not really sure how to solve this.
2) All of my tests are failing with the same error `NameError: name 'fsspec' is not defined`. I dont get this at all since the other tests using fsspec seem to succeed. Am I missing something here?

**Comment by rabernat** [comment]
> 1. I have installed pre-commit but the formatting does not seem to be identical. Not really sure how to solve this.

I'm not sure either. I learned about pre-commit from you! It would be great if someone could sort this out for rechunker.

> 2\. All of my tests are failing with the same error `NameError: name 'fsspec' is not defined`

The error I see is

```
 >       assert dst.a.data.chunksize == target_chunks_expected
  E       NameError: name 'dst' is not defined
```

File: tests/test_rechunk.py [hunk 6]
```diff
@@ -130,6 +206,80 @@ def test_rechunk_dataset(
...
+        max_mem=max_mem,
+        target_store=target_store,
+        target_options=options,
+        temp_store=temp_store,
+    )
+    assert isinstance(rechunked, api.Rechunked)
+    with dask.config.set(scheduler="single-threaded"):
+        rechunked.execute()
+
+    assert dst.a.data.chunksize == target_chunks_expected
```
Comment (rabernat) [inline tests/test_rechunk.py:276]: Looks like you have not defined `dst`. In the other test, it is defined as

```python
dst = xarray.open_zarr(target_store, decode_cf=True)
```

Comment (jbusecke) [reply tests/test_rechunk.py:276]: Well this is embarrassing. Sorry for this fail...Ill fix that right now.

**Comment by jbusecke** [comment]
Ok I figured out the linting. The problem was that this `pre-commit-config.yml` has a pretty old black version. When I was saving files things were formatted based on a different more recent version. So i just needed to make sure not to save at all after `pre-commit run --all-files`. To avoid this mismatch one could update the [black version](https://github.com/pangeo-data/rechunker/blob/73ef80b4ec9dfb980988460f7d5290076d4835e2/.pre-commit-config.yaml#L8).

I got the test running but there is a desing question that remains:

What do we want to happen when a dimension of the dataset is not specified in the `target_chunks`. For instance if we have a 2D dataset (with dims x and y). What happens when we call `rechunk(ds, ..., {'x':3})`. Should that default to the full length of y (current implementation) or preserve given chunks along y? This is a bit trickier to implement, since they could be different for different variables.

**Comment by rabernat** [comment]
> What do we want to happen when a dimension of the dataset is not specified in the `target_chunks`

I would say that the chunking should not change for that dimension.

**Comment by jbusecke** [comment]
Ok I think that would then be 100% in line with xarray, so I think that is a good way to go. Let me see how I can achieve that...

**Comment by jbusecke** [comment]
Yay. I think I got it to work. This will now preserve the chunksize (possibly different on different variables) of the existing datasets if not otherwise specified.

File: rechunker/api.py [hunk 7]
```diff
@@ -311,6 +312,43 @@ def rechunk(
...
+    """
+    group_chunks = defaultdict(list)
+    for var in ds.variables:
+        for dim in ds[var].dims:
+            if dim in target_chunks.keys() and target_chunks[dim] <= len(ds[dim]):
+                group_chunks[var].append(target_chunks[dim])
+            elif (
+                dim in target_chunks.keys()
+                and (target_chunks[dim] > len(ds[dim]) or target_chunks[dim] < 0)
+            ) or not isinstance(ds[var].data, dask.array.Array):
```
Comment (pmav99) [inline rechunker/api.py:334]: I hope I am not misreading this, but I think that If `target_chunks[dim]` is < 0 then the `if` condition on line 329 will evaluate as `True` and the `elif` condition will never be evaluated. Reversing the `if/elif` condition should fix this. Adding an explicit test for this wouldn't be a bad idea.  

That being said, I kind of dislike bundling the `isinstance()` check with the other conditions in the `elif`. I think that adding an explicit `elif` for the `isinstance()` check will make the code easier to read and easier to extend in the future if different types are supported. 

Finally, it might be worth considering whether it makes sense to put the inner loop in a separate function which would make it much easier to unit-test it. E.g.
 
``` python
def get_dim_chunk(ds, var, dim, target_chunks, existing_chunksizes):
    if not isinstance(ds[var].data, dask.array.Array):
        dim_chunk = len(ds[dim])
    elif dim in target_chunks.keys():
        if target_chunks[dim] > len(ds[dim]) or target_chunks[dim] < 0:
            dim_chunk = len(ds[dim])
        else:
            dim_chunk = target_chunks[dim]
    else:
        dim_chunk = existing_chunksizes[dim]
    return dim_chunk
```

**Comment by jbusecke** [comment]
Super sorry for dropping the ball on this. The original logic proposed by @pmav99 above was not quite giving me the output desired. But with a minimal change (I reordered the logic so that `if dim in target_chunks.keys()` is checked first), the original tests pass. I also added some specific tests for `get_dim_chunk` in the tests. 

Could you check if this looks good to you @pmav99? If so I will add docstrings and an entry to the docs.

File: tests/test_rechunk.py [hunk 8]
```diff
@@ -41,6 +41,130 @@ def test_invalid_executor():
...
+        ds=chunk_ds, target_chunks=target_chunks
+    )
+    assert expected == result
+
+
+@pytest.mark.parametrize(
+    "dask_chunks, dim, target_chunks, expected",
+    [
+        pytest.param(
+            False, "lon", dict(lon=10), 10, id="small lon chunks numpy array",
```
Comment (pmav99) [inline tests/test_rechunk.py:141]: Not too important, but semantically speaking, wouldn't `None` be a more suitable value instead of `False` for "missing/not-speficied" `dask_chunks`?

File: tests/test_rechunk.py [hunk 9]
```diff
@@ -41,6 +41,130 @@ def test_invalid_executor():
...
+            365,
+            id="negative time chunks dask array",
+        ),
+        pytest.param(
+            {"time": 5}, "time", dict(), 5, id="preserve small time chunks dask array",
+        ),
+        pytest.param(
+            False, "lon", dict(), 360, id="preserve full length chunk numpy array",
+        ),
+    ],
```
Comment (pmav99) [inline tests/test_rechunk.py:159]: This is purely a personal preference, so feel free to ignore it. Exceeding the line length limit does not unconditionally mean that the code is less readable. `pytest.param()` lines tend to often be longer than what makes sense for normal code. Therefore I usually keep them in a single line and ask `black` to ignore the relevant lines by adding `fmt: off` and `fmt: on` comments. E.g.

``` python
@pytest.mark.parametrize(
    "dask_chunks, dim, target_chunks, expected",
    [
        # fmt: off
        pytest.param(False, "lon", dict(lon=10), 10, id="small lon chunks numpy array",),
        pytest.param(False, "lon", dict(lon=10), 10, id="small lon chunks dask array",),
        pytest.param(False, "time", dict(time=400), 365, id="time chunks exceed len",),
        pytest.param({"time": 1}, "time", dict(time=-1), 365, id="negative time chunks dask array",),
        # fmt: on
    ]
)
def test_get_dim_chunk(dask_chunks, chunk_ds, dim, target_chunks, expected):
    ...
```

**Review by pmav99 — APPROVED** [review]
I am afraid I don't have the time to test the changes on real datasets. Overall it looks great. I 've made a couple of minor/stylistic remarks inline. Feel free to ignore them.

The only thing that I am not really fond of is the length of `test_rechunk_dataset_dimchunks` (60 lines and, probably, counting). Ideally the dataset creation should be done in a function or in a pytest fixture. Nevertheless, at the moment, both session scoped fixtures **and** functions are being used for this purpose (in the particular module). So it is not easy to make a choice. Therefore I would suggest to keep this as it is, merge the code and perhaps refactor the test suite in the future.

**Comment by jbusecke** [comment]
> The only thing that I am not really fond of is the length of test_rechunk_dataset_dimchunks (60 lines and, probably, counting). Ideally the dataset creation should be done in a function or in a pytest fixture. Nevertheless, at the moment, both session scoped fixtures and functions are being used for this purpose (in the particular module). So it is not easy to make a choice. Therefore I would suggest to keep this as it is, merge the code and perhaps refactor the test suite in the future.

That makes sense, but since this was basically just adopted from the existing tests, I think your suggestions to refactor at a later time would be better (rather than mixing new features with changes in existing tests?).

File: tests/test_rechunk.py (hunk 9 above)
Comment (jbusecke) [reply tests/test_rechunk.py:159]: I didnt even know this was an option. Happy to implement.

File: tests/test_rechunk.py (hunk 8 above)
Comment (jbusecke) [reply tests/test_rechunk.py]: Sure thing.

**Comment by rabernat** [comment]
I concur with the comments about the need to refactor the tests. This is an example of maintenance work that would really benefit rechunker but that the original devs have not managed to find time for. It would be great if @jbusecke and @pmav99 were willing to get more involved and take this on eventually.

I am willing to merge this as is, but I will open an issue about the testing refactor.

File: tests/test_rechunk.py [hunk 10]
```diff
@@ -41,6 +41,117 @@ def test_invalid_executor():
...
+
+
+@pytest.mark.parametrize(
+    "target_chunks,expected",
+    [
+        pytest.param(
+            dict(lon=10),
+            dict(aaa=(10, 180, 365), lon=(10,), lat=(180,), time=(365,)),
+            id="just lon chunk",
+        ),
```
Comment (rabernat) [inline tests/test_rechunk.py]: I did not know about this syntax. Thanks for teaching me something new! 🤓

File: tests/test_rechunk.py [hunk 11]
```diff
@@ -131,6 +242,86 @@ def test_rechunk_dataset(
     assert ds.attrs == dst.attrs
 
 
+@pytest.mark.parametrize("shape", [(100, 50)])
+@pytest.mark.parametrize("source_chunks", [(10, 50), (100, 5)])
+@pytest.mark.parametrize(
+    "target_chunks",
+    [
+        {"x": 20},  # ? Should this rechunk y? Probably not...
+        {"x": 20, "y": 1e5},
```
Comment (rabernat) [inline tests/test_rechunk.py:200]:
```suggestion
        {"x": 20, "y": 100_000},
```

Better to use integer chunks

File: tests/test_rechunk.py [hunk 12]
```diff
@@ -131,6 +242,86 @@ def test_rechunk_dataset(
...
+    ],
+)
+@pytest.mark.parametrize("max_mem", ["10MB"])
+@pytest.mark.parametrize("target_store", ["target.zarr", "mapper.target.zarr"])
+@pytest.mark.parametrize("temp_store", ["temp.zarr", "mapper.temp.zarr"])
+def test_rechunk_dataset_dimchunks(
+    tmp_path, shape, source_chunks, target_chunks, max_mem, target_store, temp_store,
+):
+    if target_store.startswith("mapper"):
+        fsspec = pytest.importorskip("fsspec")
```
Comment (rabernat) [inline tests/test_rechunk.py:211]: Is fsspec in our test env? I'm confused why we are mixing in fsspec mapper stuff with this new feature about chunking.

File: tests/test_rechunk.py [hunk 13]
```diff
@@ -131,6 +242,86 @@ def test_rechunk_dataset(
...
+        target_store = fsspec.get_mapper(str(tmp_path) + target_store)
+        temp_store = fsspec.get_mapper(str(tmp_path) + temp_store)
+    else:
+        target_store = str(tmp_path / target_store)
+        temp_store = str(tmp_path / temp_store)
+
+    a = numpy.arange(numpy.prod(shape)).reshape(shape).astype("f4")
+    a[-1] = numpy.nan
```
Comment (rabernat) [inline tests/test_rechunk.py:219]: What is this line for?

File: tests/test_rechunk.py [hunk 14]
```diff
@@ -131,6 +242,86 @@ def test_rechunk_dataset(
...
+            a=xarray.DataArray(
+                a, dims=["x", "y"], attrs={"a1": 1, "a2": [1, 2, 3], "a3": "x"}
+            ),
+            b=xarray.DataArray(numpy.ones(shape[0]), dims=["x"]),
+            c=xarray.DataArray(numpy.ones(shape[1]), dims=["y"]),
+        ),
+        coords=dict(
+            cx=xarray.DataArray(numpy.ones(shape[0]), dims=["x"]),
+            cy=xarray.DataArray(numpy.ones(shape[1]), dims=["y"]),
+        ),
```
Comment (rabernat) [inline tests/test_rechunk.py:231]: There is a more concise way to construct an Xarray dataset. See http://xarray.pydata.org/en/stable/user-guide/data-structures.html#creating-a-dataset

Would prefer not to be creating DataArrays here.

File: tests/test_rechunk.py [hunk 15]
```diff
@@ -131,6 +242,86 @@ def test_rechunk_dataset(
...
+    )
+    ds = ds.chunk(chunks=dict(zip(["x", "y"], source_chunks)))
+    options = dict(
+        a=dict(
+            compressor=zarr.Blosc(cname="zstd"),
+            dtype="int32",
+            scale_factor=0.1,
+            _FillValue=-9999,
+        )
+    )
```
Comment (rabernat) [inline tests/test_rechunk.py:242]: Does this test really need options? This seems orthogonal to the feature being tested and makes the test longer / harder to read / maintain.

**Review by rabernat — CHANGES_REQUESTED** [review]
Just had a full read through. This looks awesome! 🎉 

I have a few minor nit picks.

File: tests/test_rechunk.py (hunk 13 above)
Comment (jbusecke) [reply tests/test_rechunk.py:219]: No clue. Literally just copied that test from the existing ones.

**Comment by jbusecke** [comment]
I see your point @rabernat , but the problem for me is that I do not know a lot of the internals here, so it will take me some time to figure this out. 

In the meantime, I have a more important design question regarding this feature. I was just looking through the tutorial in the [docs](https://rechunker.readthedocs.io/en/latest/tutorial.html#Rechunk-a-Group) and I am wondering if we actually want to chunk dimension arrays? I believe the current code will chunk e.g. a 1D dimension (`x`) array the same way as an ND array which has multiple dimensions (`x`, `y`,`z`). I am not sure if that is desireable or not.

**Comment by rabernat** [comment]
> I am wondering if we actually want to chunk dimension arrays?

The original version of rechunker only operated on Zarr arrays / groups, without no option for Xarray inputs. That was added after. So we really don't distinguish between dimension and non-dimension coordinates (or between coordinates and data variables at all). This is what we say about rechunking Zarr groups

> For a group of arrays, a dict is required. The keys correspond to array names. The values are `target_chunks` arguments for the array. For example, `{'foo': (20, 10), 'bar': {'x': 3, 'y': 5}, 'baz': None}`. All arrays you want to rechunk must be explicitly named. Arrays that are not present in the `target_chunks` dict will be ignored.

We don't really specify the behavior for Xarray datasets clearly. Which means it is yet to be defined.

Looking more closely at the code, I am starting to think we need to split the current high-level API, which basically only exposes the `rechunk` function, into several different functions:
- `rechunk_zarr_array`
- `rechunk_zarr_group`
- `rechunk_dask_array`
- `rechunk_xarray_dataset`

...and deprecate `rechunk`. That would allow us to have different options and default that are appropriate to each scenario. It would also remove a ton of ugly if / else blocks from the code.

Obviously we don't want to do that in this PR. In the meantime, I would say that no special treatment of dimension coordinates is needed here.

File: tests/test_rechunk.py (hunk 9 above)
Comment (jbusecke) [reply tests/test_rechunk.py]: @pmav99 this seems to work with black, but not flake8. Is there a similar workaround? Otherwise I would probably return this to the original state to move this along if that is ok?

**Comment by jbusecke** [comment]
Sorry for the long pause on this. I am back from vacation and eager to use this feature. 
I have trimmed the test parameterizations in the test specific to the new chunking method to a minimum, and made some comments reflecting @rabernat s comments above. Do you think this is sufficient on the code level? Ill go ahead and add a little example for the docs now.

**Comment by jbusecke** [comment]
Ok I added an example but encountered a bunch of issues with the tutorial that make it quite ugly as is. I wonder if we should merge the code changes here as is and move the docs to another PR (happy to work on that), which updates the docs and addresses #103, #104, and #105?

File: tests/test_rechunk.py (hunk 13 above)
Comment (rabernat) [reply tests/test_rechunk.py:254]: Rather than copying this somewhat mysterious logic into a new test, let's at least move it into a fixture so we can eventually fix it once and for all.

File: tests/test_rechunk.py (hunk 14 above)
Comment (rabernat) [reply tests/test_rechunk.py:266]: Please use a different syntax for creating DataArrays and / or move this too a fixture so as not to accumulate further technical debt in the test suite.

File: tests/test_rechunk.py (hunk 15 above)
Comment (rabernat) [reply tests/test_rechunk.py:277]: Please address this review comment.

**Review by rabernat — CHANGES_REQUESTED** [review]
Overall this looks really good.

However, I am concerned we are accumulating even more technical debt by copy-pasting test setup that is hard to understand. Let's at least move this to a fixture so it only appears in one place.

Then LGTM.

**Comment by jbusecke** [comment]
> Ok I added an example but encountered a bunch of issues with the tutorial that make it quite ugly as is. I wonder if we should merge the code changes here as is and move the docs to another PR (happy to work on that), which updates the docs and addresses #103, #104, and #105?

Oh I just noticed that I accidentally included the new tutorial (with a bunch of failures) into this PR. Any opinion on how to solve this best?

**Comment by rabernat** [comment]
Oops. Sorry for not catching that. Could you make a new PR to fix that (and any other issues you found).

**Comment by jbusecke** [comment]
Yep, will do. Just waiting for #106 to be merged.
